controller/src/main/java/io/pravega/controller/server/ControllerServiceMain.java

```diff
@@ -120,6 +120,9 @@ protected void run() throws Exception {
                 notifyServiceStateChange(ServiceState.STARTING);
                 starter.startAsync();
 
+                sessionExpiryFuture.whenComplete((r, e) -> {
+                     starter.notifySessionExpiration();
```

Comment:
short answer, no we cannot move it to inside that `if block` because `ControllerServiceStarter` object is not present at that time. 

this is a callback registered on the session expiration future. and in the callback we send a notification to the starter object.

Reply:
I want to understand what the benefit of that suggestion would be? It would only be cosmetic. 

Following is what we do in current code:
1. we create curator client 
2. then we register for session expiration callback first thing before even attempting to perform any queries on the zookeeper session. In the callback we complete sessionExpirationFuture. 
3. we create the `starter` object using the storeClient which is a wrapper over the curator client. 
4. we register a callback on `sessionExpirationFuture` to call starter.notifySessionExpiration. 
5. we start the starter. 

we can do `3` before `2` and merge `4` into `2`. But what will we achieve with that change?

Reply:
> I want to understand what the benefit of that suggestion would be?

It's code readability. We complete `sessionExpiryFuture` in line 112, it makes sense that we notify the starter right after completing that future. When I checked this code change the first time, I didn't immediately understand why we had a `sessionExpiryFuture.whenComplete` in the middle of the flow of that method.